Client/Frontend/Browser/BrowserViewController/OpenSearch/BrowserViewController+OpenSearch.swift

```diff
@@ -0,0 +1,217 @@
+        } else {
+            self.customSearchEngineButton.state = .enabled
+        }
+        
+        /*
+         This is how we access hidden views in the WKContentView
+         Using the public headers we can find the keyboard accessoryView which is not usually available.
+         Specific values here are from the WKContentView headers.
+         https://github.com/JaviSoto/iOS9-Runtime-Headers/blob/master/Frameworks/WebKit.framework/WKContentView.h
+         */
+        guard let webContentView = UIView.findSubViewWithFirstResponder(webView) else {
+            /*
+             In some cases the URL bar can trigger the keyboard notification. In that case the webview isnt the first responder
+             and a search button should not be added.
+             */
+            return
+        }
+        
+        let argumentNextItem: [Any] = ["_n", "extI", "tem"]
+        let argumentView: [Any] = ["v", "ie", "w"]
+        
+        let valueKeyNextItem = argumentNextItem.compactMap { $0 as? String }.joined()
+        let valueKeyView = argumentView.compactMap { $0 as? String }.joined()
+
+        guard let input = webContentView.perform(#selector(getter: UIResponder.inputAccessoryView)),
+              let inputView = input.takeUnretainedValue() as? UIInputView,
+              let nextButton = inputView.value(forKey: valueKeyNextItem) as? UIBarButtonItem,
+              let nextButtonView = nextButton.value(forKey: valueKeyView) as? UIView else {
+            return
+        }
+        
+        inputView.addSubview(customSearchEngineButton)
+        
+        customSearchEngineButton.snp.remakeConstraints { make in
+            make.leading.equalTo(nextButtonView.snp.trailing).offset(20)
+            make.width.equalTo(inputView.snp.height)
+            make.top.equalTo(nextButtonView.snp.top)
+            make.height.equalTo(inputView.snp.height)
+        }
+    }
+
+    @objc func addCustomSearchEngineForFocusedElement() {
+        guard var referenceURLString = openSearchEngine?.reference,
+              let title = openSearchEngine?.title,
+              var referenceURL = URL(string: referenceURLString),
+              let faviconURLString = self.tabManager.selectedTab?.displayFavicon?.url,
+              let iconURL = URL(string: faviconURLString) else {
+            let alert = ThirdPartySearchAlerts.failedToAddThirdPartySearch()
+            present(alert, animated: true, completion: nil)
+            return
+        }
+                
+        guard let scheme = tabManager.selectedTab?.webView?.url?.scheme,
+              let host = tabManager.selectedTab?.webView?.url?.host else {
+            log.error("Selected Tab doesn't have URL")
+            return
+        }
+        
+        while referenceURLString.hasPrefix("/") {
+            referenceURLString.remove(at: referenceURLString.startIndex)
+        }
+        
+        let constructedReferenceURLString = "\(scheme)://\(host)/\(referenceURLString)"
```

Comment:
I believe it might be better this way to understand what is being passed dd and how the reference URL is constructed. We are guarding the scheme and host we receive from active tab.